src/coreclr/src/jit/compiler.h

```diff
@@ -3457,10 +3457,12 @@ class Compiler
 
         // We make local variable SIMD12 types 16 bytes instead of just 12. lvSize()
         // already does this calculation. However, we also need to prevent mapping types if the var is a
-        // dependently promoted struct field, which must remain its exact size within its parent struct.
+        // dependently promoted struct field, which must remain its exact size within its parent struct,
+        // unless it is the only field.
         // However, we don't know this until late, so we may have already pretended the field is bigger
         // before that.
-        if ((varDsc->lvSize() == 16) && !lvaIsFieldOfDependentlyPromotedStruct(varDsc))
+        if ((varDsc->lvSize() == 16) &&
```

Comment:
can we check that parent exact size is 16? To protect from cases like:
```
[StructLayout(LayoutKind.Explicit, Pack = 1)]
public struct A {
  [StructLayout(LayoutKind.Explicit, Pack = 1)]
  public struct B {
    SIMD16 c; // can't be written as 16 bytes.
  }
}
```
write now we won't generate a `LCL_FLD` node to access `c`, but it could change soon.

Reply:
Should that be `SIMD12 c;`?

Reply:
Have you verified that a struct such as that will occupy only 12 bytes on the stack? I would be surprised, as I would expect it to always get 16. In any case, it is an inexpensive check to add.

Reply:
Actually, I take that back - `lvExactSize` will be 12 so we will miss the optimization in all cases. But if we are actually packing such a type, e.g. on a 32-bit architecture, `lvSize()` will return 12. So I believe that the check is correct.

Reply:
I'm pretty sure that we rely in many places on `lvSize()` returning the size that the variable is allocated on the stack.

Reply:
You can see here: https://github.com/dotnet/runtime/blob/master/src/coreclr/src/jit/lclvars.cpp#L6351 that we allocate `lvSize()` on the stack.

Reply:
If `lvSize() == 16` means that it takes 16 bytes on the stack, doesn't it mean that if we map it to `SIMD16` it won't overlap with any parent fields that it was not overlapping before?
If the parent struct was promoted it means there were no overlapping fields, so it is not clear for me why we need to check anything except `if (varDsc->lvSize() == 16)`

Reply:
The reason we need to make the other checks is that if it has multiple fields, it is possible (in particular on 32-bit targets) that it would have a 32-bit field following it. This could even happen on 64-bit targets with explicit layout.

Reply:
> it is possible (in particular on 32-bit targets) that it would have a 32-bit field following it. 

do I understand that correctly: 
```
struct C {   // size 16 on x86?
  SIMD12 a;  // offset 0, lvSize() == 16?, lvExactSize == 12
  int b;     // offset 12? , size == 4
}
```
how do we allocate the second field starting at 12 after we passed `16` to [lvaAllocLocalAndSetVirtualOffset(lclNum, lvaLclSize(lclNum), stkOffs);](https://github.com/dotnet/runtime/blob/c9f917052fc091014ef0f311b6f79a812bbfbd38/src/coreclr/src/jit/lclvars.cpp#L6352)

Reply:
Sorry I missed this comment earlier (I'm still struggling to get green CI testing on this!). We special-case locals of `SIMD12` type on 32-bit targets so that they occupy a full 16 bytes, where normally it would only be 12, since that's an even multiple of `TARGET_POINTER_SIZE`. However, if it is a member of another struct we can't do that (it would violate the struct packing rules). So on a 32-bit target the size of `C.a` would be 12.

Reply:
I've backed out this change - it's not clear from the comment here (and I hadn't remembered) that `lvSize()` always returns the "fake" size of `TYP_SIMD12`, even if it may later become a "dependently promoted" local. This was causing the RayTracer benchmark to fail, because the local is 12 bytes but we were storing 16.